### PR TITLE
[Exporter.Geneva] Avoid allocation for serializing scopes

### PR DESCRIPTION
**Pull request by utpilla**
## Changes
- Avoid allocation for serializing scopes

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes

### main branch:

|                    Method | IncludeFormattedMessage |     Mean |   Error |  StdDev |   Gen0 | Allocated |
|-------------------------- |------------------------ |---------:|--------:|--------:|-------:|----------:|
| LoggerWithMessageTemplate |                   False | 871.7 ns | 3.08 ns | 2.73 ns | 0.0324 |     208 B |
| LoggerWithDirectLoggerAPI |                   False | 818.1 ns | 4.09 ns | 3.83 ns | 0.0544 |     344 B |
| LoggerWithSourceGenerator |                   False | 857.5 ns | 2.14 ns | 1.90 ns | 0.0267 |     168 B |
|        SerializeLogRecord |                   False | 660.2 ns | 6.22 ns | 5.82 ns | 0.0162 |     104 B |
|                    Export |                   False | 694.2 ns | 2.49 ns | 2.21 ns | 0.0162 |     104 B |
| LoggerWithMessageTemplate |                    True | 894.9 ns | 1.94 ns | 1.72 ns | 0.0324 |     208 B |
| LoggerWithDirectLoggerAPI |                    True | 811.7 ns | 1.65 ns | 1.54 ns | 0.0544 |     344 B |
| LoggerWithSourceGenerator |                    True | 870.4 ns | 2.01 ns | 1.88 ns | 0.0267 |     168 B |
|        SerializeLogRecord |                    True | 659.9 ns | 5.80 ns | 5.43 ns | 0.0162 |     104 B |
|                    Export |                    True | 694.3 ns | 2.32 ns | 2.17 ns | 0.0162 |     104 B |

### With this PR:

|                    Method | IncludeFormattedMessage |     Mean |   Error |  StdDev |   Gen0 | Allocated |
|-------------------------- |------------------------ |---------:|--------:|--------:|-------:|----------:|
| LoggerWithMessageTemplate |                   False | 832.5 ns | 5.99 ns | 5.00 ns | 0.0162 |     104 B |
| LoggerWithDirectLoggerAPI |                   False | 766.2 ns | 3.85 ns | 3.60 ns | 0.0381 |     240 B |
| LoggerWithSourceGenerator |                   False | 815.3 ns | 2.89 ns | 2.41 ns | 0.0095 |      64 B |
|        SerializeLogRecord |                   False | 582.3 ns | 0.81 ns | 0.72 ns |      - |         - |
|                    Export |                   False | 646.0 ns | 1.10 ns | 0.86 ns |      - |         - |
| LoggerWithMessageTemplate |                    True | 847.7 ns | 5.56 ns | 5.20 ns | 0.0162 |     104 B |
| LoggerWithDirectLoggerAPI |                    True | 762.5 ns | 2.72 ns | 2.41 ns | 0.0381 |     240 B |
| LoggerWithSourceGenerator |                    True | 816.6 ns | 2.79 ns | 2.47 ns | 0.0095 |      64 B |
|        SerializeLogRecord |                    True | 586.3 ns | 1.80 ns | 1.69 ns |      - |         - |
|                    Export |                    True | 659.5 ns | 6.00 ns | 5.61 ns |      - |         - |

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs [hunk 1]
```diff
@@ -399,24 +397,15 @@ void ProcessScopeForIndividualColumns(LogRecordScope scope, object state)
...
-            {
-                foreach (KeyValuePair<string, object> scopeItem in scope)
-                {
-                    if (string.IsNullOrEmpty(scopeItem.Key) || scopeItem.Key == "{OriginalFormat}")
-                    {
-                        continue;
-                    }
+            // Prepare state for scopes
+            dataForScopes.Cursor = cursor;
+            dataForScopes.EnvPropertiesCount = envPropertiesCount;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs:402]: Do we need to reset `HasEnvProperties` here? Looks like the scope delegate doesn't always set it.

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs:402]: Maybe also `FieldsCount`

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs:402]: Maybe up around Line 367.

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs [hunk 2]
```diff
@@ -33,11 +33,12 @@ internal sealed class MsgPackLogExporter : MsgPackExporter, IDisposable
     private readonly IReadOnlyDictionary<string, object> m_customFields;
 
     private readonly ExceptionStackExportMode m_exportExceptionStack;
+    private readonly ExporterStateForScopes exporterStateForScopes;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs:36]: nit: `m_exporterStateForScopes` to be consistent with the other fields.

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs [hunk 3]
```diff
@@ -127,6 +128,8 @@ public MsgPackLogExporter(GenevaExporterOptions options)
             this.m_customFields = customFields;
         }
 
+        this.exporterStateForScopes = new ExporterStateForScopes() { CustomFields = this.m_customFields };
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs:131]: nit: Just reading the diff, it doesn't seem like `ExporterStateForScopes` is needed. You could do something more or less like this...

```csharp
class MsgPackLogExporter
{
    private readonly ThreadLocal<SerializationDataForScopes> m_serializationData = new(() => null);

    private static readonly Action<LogRecordScope, MsgPackLogExporter> ProcessScopeForIndividualColumns = (scope, state) =>
    {
        var stateData = state.m_serializationData.Value;
        var customFields = state.m_customFields;

        // Do work
    }
}
```

Basically use `MsgPackLogExporter` as the state type.

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs (hunk 1 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs:403]: Updated.

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs (hunk 2 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs]: Updated. I think we should update the field names to use Pascal case instead though. 😄

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs (hunk 3 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackLogExporter.cs]: Great suggestion! Updated.

**Review by CodeBlanch — APPROVED** [review]

**Review by CodeBlanch — APPROVED** [review]

**Pull request merged**